File: assignments/08-strings/src/Hard/Q8.py
def makeLargestSpecial(s:str) -> str:
    count = i = 0

    res=[]

    for j,v in enumerate(s):
        count = count+1 if v == '1' else count-1

        if count == 0:
            res.append('1' + makeLargestSpecial(s[i+1:j]) + '0')
            i=j+1
    return ''.join(sorted(res)[::-1])


print(makeLargestSpecial("11011000"))

File: assignments/08-strings/src/Hard/Q8.java
package Hard;

import java.util.Collections;
import java.util.List;
import java.util.ArrayList;

public class Q8 {
    
    public String makeLargestSpecial(String S) {
        int count = 0, i = 0;
        List<String> res = new ArrayList<String>();
        for (int j = 0; j < S.length(); ++j) {
          if (S.charAt(j) == '1') count++;
          else count--;
          if (count == 0) {
            res.add('1' + makeLargestSpecial(S.substring(i + 1, j)) + '0');
            i = j + 1;
          }
        }
        Collections.sort(res, Collections.reverseOrder());
        return String.join("", res);
    }

    public static void main(String[] args) {
        System.out.println(new Q8().makeLargestString("11011000"));
    }
}
